src/field_theory/separable_degree.lean

```diff
@@ -0,0 +1,179 @@
+`g'` agree, then either their degrees are the same, or one of them is not separable. -/
+lemma contraction_degree_eq_aux [hq : fact q.prime] [hF : char_p F q]
+  (g g' : polynomial F) (m m' : ℕ)
+  (h_expand : expand F (q^m) g = expand F (q^m') g')
+  (h : m < m') :
+  (g.nat_degree =  g'.nat_degree) ∨ ¬g.separable :=
+begin
+  suffices : g.separable → g.nat_degree = g'.nat_degree,
+  from (not_or_of_imp this).symm,
+
+  cases exists_eq_add_nonneg_of_lt m m' h with s hs,
+  intro g_sep,
+
+  rw [hs.2, pow_add, expand_mul] at h_expand,
+
+  have r := expand_injective (pow_pos hq.1.pos m) h_expand,
+  rw r at g_sep,
+
+  cases (is_unit_or_eq_zero_of_separable_expand q s g_sep) with g'_is_unit s_zero,
+  { rw [r, nat_degree_expand (q^s) g',
+      nat_degree_eq_of_degree_eq_some (degree_eq_zero_of_is_unit g'_is_unit), zero_mul ] },
+  { apply false.elim,
+    exact (ne_of_lt hs.1).symm s_zero },
+end
+
+/-- If two expansions (along the positive characteristic) of two polynomials `g` and `g'` agree,
+then they have the same degree or one of them is inseparable. -/
+theorem contraction_degree_eq_or_insep
+  [hq : fact q.prime] [char_p F q]
+  (g g' : polynomial F) (m m' : ℕ)
+  (h_expand : expand F (q^m) g = expand F (q^m') g') :
+  (g.nat_degree = g'.nat_degree) ∨ ¬g.separable ∨ ¬g'.separable :=
+begin
+  by_cases h : m = m',
+  { apply or.inl,
+    -- if `m = m'` then we show `g.nat_degree = g'.nat_degree` by unfolding the definitions
+    rw h at h_expand,
+    have expand_deg : ((expand F (q ^ m')) g).nat_degree =
+      (expand F (q ^ m') g').nat_degree, by rw h_expand,
+    rw [nat_degree_expand (q^m') g, nat_degree_expand (q^m') g'] at expand_deg,
+    apply nat.eq_of_mul_eq_mul_left (pow_pos hq.1.pos m'),
+    rw [mul_comm] at expand_deg, rw expand_deg, rw [mul_comm] },
+  { cases ne.lt_or_lt h,
+    { rw ←or_assoc, -- if `m < m'`, we show that the degrees agree or that `g'` is not separable
+      apply or.inl,
+      apply contraction_degree_eq_aux q g g' m m' h_expand h_1 },
+    { rw ←or.left_comm, -- if `m' < m`, we show that the degrees agree or that `g` is not separable
+      apply or.inr,
+      cases contraction_degree_eq_aux q g' g m' m h_expand.symm h_1,
+      exact or.inl h_2.symm,
+      exact or.inr h_2 } }
+end
+
+/-- The separable degree is unique. -/
+theorem separable_degree_eq {n n' : ℕ} [hF : exp_char F q] :
+  separable_degree q hf = n → separable_degree q hf = n' → n = n' :=
+λ h, λ h',
+begin
+  casesI hF,
+  { rw separable_degree_eq_degree hf at h h',
+    rw h at h',
+    exact h' },
+  { let g := classical.some hf,
+    cases (classical.some_spec hf).2 with m hm,
+
+    let g' := classical.some hf,
+    cases (classical.some_spec hf).2 with m' hm',
+
+    suffices : g.nat_degree = g'.nat_degree ∨ ¬g.separable ∨ ¬g'.separable, by tidy,
+
+    haveI : fact q.prime := fact_iff.2 hF_hprime,
+    apply contraction_degree_eq_or_insep q g g' m m',
+    rw [hm, hm'] }
+end
```

Comment:
I'm pretty sure this doesn't state what you want it to state, as the proof is just "equality is transitive":
```suggestion
/-- The separable degree is unique. -/
theorem separable_degree_eq {n n' : ℕ} [hF : exp_char F q] :
  separable_degree q hf = n → separable_degree q hf = n' → n = n' :=
λ h, h.symm.trans
```

Reply:
This is now fixed.